cache: fix per-response caching and switch to pointers for requests and responses

joshua-rutherford opened the pull request:
Adds test assertions for caching within the `RawMessage.GetDiscoveryResponse` method and ensures the caching is occurring as intended.

jyotimahapatra commented on pkg/cache/v2/cache.go:
```diff
@@ -104,55 +99,55 @@ var _ Response = &PassthroughResponse{}
+	if r.marshaledResponse == nil {
 
-	marshaledResources := make([]*any.Any, len(r.Resources))
+		marshaledResources := make([]*any.Any, len(r.Resources))
```
nit: i think there's extra indentation on lines 106-116

jyotimahapatra commented:
LGTM!

alecholmez commented:
I think the only thing I saw was I changed this:
https://github.com/envoyproxy/go-control-plane/blob/ad54bfba2b9328feade0962fb8b1ab4a37e6b1a3/pkg/cache/v2/cache.go#L27-L28
for the delta counterparts in my PR. As the code stands it copies a lock in `CreateWatch` and `SetSnapshot`. I just changed that line to a pointer and fixed the resulting errors and the lock lint complaints went away

kyessenov commented:
Have you tried running with a race detector (go -race)? I think golang does not guarantee atomicity of pointer reads and writes unfortunately. You have to use something clunky like https://golang.org/pkg/sync/atomic/#Value.

kyessenov commented:
I think this is fine to merge. I am just wondering if golang catches what it deems to be unsafe (even though I think this works fine on x86).

kyessenov commented:
Yup, I think DATA RACE is triggered on 1.14 compiler.

jyotimahapatra commented:
>I think golang does not guarantee atomicity of pointer reads and writes
@kyessenov I think even though atomicity is not guarenteed, because of using just a single pointer as source of data, we might end up spending extra cycles, but this will be functionally correct. WDYT?

kyessenov commented:
It's not clear from the spec. I can imagine pointer reads and writes take two instructions on some odd hardware (compressed pointers, or what not, 64 bit pointers on 32 bit machine), and they can interleave.

joshua-rutherford commented:
as per the comments from @alecholmez I've refactored to remove the `sync.Mutex` being passed by value therefore not really be in a mutex at all. 

on the points above, if a pointer write is a multi-instruction operation I'd prefer the `Value` usage as suggested.

joshua-rutherford commented:
hmm, will work the failing integration test tonight.

kyessenov commented:
On AMD64, atomic stuff is just a no-op (see https://golang.org/src/runtime/internal/atomic/atomic_amd64.go). But ARM stuff is way more complicated. I would like to think this code will run on ARM some day.

joshua-rutherford commented:
so I fixed the integration test failures (a missed mutex copy) and also took an integration test fix from #335 that was causing erroneous failures on my box.

i've also added a separate commit that modifies the implementation to ensure proper behavior on platforms where pointer access and assignment may not be atomic. 

i'm fine with either implementation so wanted to put them both out there for review as both work as expected in tests.

jyotimahapatra commented:
LGTM @joshua-rutherford .
I tried to leverage the race detector https://golang.org/doc/articles/race_detector.html#Usage by running `go test -race -v pkg/cache/v2/cache_test.go`, but it did not generate a report. 
Can you run the race detector before/after your change and see what comes up?

joshua-rutherford commented:
@jyotimahapatra, i get nothing for master, without atomic and with atomic.

```
➜  go-control-plane git:(caching) git reset --hard upstream/master                                            
HEAD is now at 9c7e2b25 Mirrored from envoyproxy/envoy @ 90a97c7ef3d67e213980db6b0ebeff284a938837
➜  go-control-plane git:(caching) ./build/run_docker.sh go test -race -v pkg/cache/v2/cache_test.go
go: downloading github.com/golang/protobuf v1.4.2
go: downloading google.golang.org/protobuf v1.23.0
go: extracting github.com/golang/protobuf v1.4.2
go: downloading google.golang.org/grpc v1.27.0
go: extracting google.golang.org/protobuf v1.23.0
go: downloading github.com/cncf/udpa/go v0.0.0-20200629203442-efcf912fb354
go: extracting github.com/cncf/udpa/go v0.0.0-20200629203442-efcf912fb354
go: extracting google.golang.org/grpc v1.27.0
go: finding github.com/golang/protobuf v1.4.2
go: finding google.golang.org/protobuf v1.23.0
go: finding github.com/cncf/udpa/go v0.0.0-20200629203442-efcf912fb354
go: finding google.golang.org/grpc v1.27.0
=== RUN   TestResponseGetDiscoveryResponse
--- PASS: TestResponseGetDiscoveryResponse (0.00s)
=== RUN   TestPassthroughResponseGetDiscoveryResponse
--- PASS: TestPassthroughResponseGetDiscoveryResponse (0.00s)
PASS
ok  	command-line-arguments	1.062s
➜  go-control-plane git:(caching) git reset --hard 830cf0f30b450bfb2cfab8118070c3cf1ec5aa1c
HEAD is now at 830cf0f3 cache: fix per response caching and switch to pointers for response
➜  go-control-plane git:(caching) ./build/run_docker.sh go test -race -v pkg/cache/v2/cache_test.go
go: downloading github.com/golang/protobuf v1.4.2
go: downloading github.com/cncf/udpa/go v0.0.0-20200629203442-efcf912fb354
go: downloading google.golang.org/protobuf v1.23.0
go: extracting github.com/golang/protobuf v1.4.2
go: extracting github.com/cncf/udpa/go v0.0.0-20200629203442-efcf912fb354
go: downloading google.golang.org/grpc v1.27.0
go: extracting google.golang.org/protobuf v1.23.0
go: extracting google.golang.org/grpc v1.27.0
go: finding github.com/golang/protobuf v1.4.2
go: finding google.golang.org/protobuf v1.23.0
go: finding github.com/cncf/udpa/go v0.0.0-20200629203442-efcf912fb354
go: finding google.golang.org/grpc v1.27.0
=== RUN   TestResponseGetDiscoveryResponse
--- PASS: TestResponseGetDiscoveryResponse (0.00s)
=== RUN   TestPassthroughResponseGetDiscoveryResponse
--- PASS: TestPassthroughResponseGetDiscoveryResponse (0.00s)
PASS
ok  	command-line-arguments	1.061s
➜  go-control-plane git:(caching) git reset --hard 830cf0f30b450bfb2cfab8118070c3cf1ec5aa1c        
HEAD is now at 830cf0f3 cache: fix per response caching and switch to pointers for response
➜  go-control-plane git:(caching) ./build/run_docker.sh go test -race -v pkg/cache/v2/cache_test.go
go: downloading github.com/golang/protobuf v1.4.2
go: downloading github.com/cncf/udpa/go v0.0.0-20200629203442-efcf912fb354
go: downloading google.golang.org/protobuf v1.23.0
go: downloading google.golang.org/grpc v1.27.0
go: extracting github.com/golang/protobuf v1.4.2
go: extracting github.com/cncf/udpa/go v0.0.0-20200629203442-efcf912fb354
go: extracting google.golang.org/grpc v1.27.0
go: extracting google.golang.org/protobuf v1.23.0
go: finding github.com/golang/protobuf v1.4.2
go: finding google.golang.org/protobuf v1.23.0
go: finding github.com/cncf/udpa/go v0.0.0-20200629203442-efcf912fb354
go: finding google.golang.org/grpc v1.27.0
=== RUN   TestResponseGetDiscoveryResponse
--- PASS: TestResponseGetDiscoveryResponse (0.00s)
=== RUN   TestPassthroughResponseGetDiscoveryResponse
--- PASS: TestPassthroughResponseGetDiscoveryResponse (0.00s)
PASS
ok  	command-line-arguments	1.059s
```

kyessenov commented:
We probably don't have a test where we concurrently try to cache raw response. It's easy to reproduce DATA RACE with two concurrent go routine that try to read/write same pointer. In reality, this atomic is just an unsafe pointer on x86 so only incurs type casting overhead.
